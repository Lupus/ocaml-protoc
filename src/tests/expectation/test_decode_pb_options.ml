module E = Pb_exception
module Pt = Pb_parsing_parse_tree

let process_field_options ppf field =
  let options =
    field.Pt.field_options |> Pb_typing_validation.compile_options
  in
  if options <> Pb_option.empty then (
    let parsed =
      Pb_option.get_ext options "validate.rules"
      |> Option.map Validate.decode_pb_options_field_rules
    in
    Format.fprintf ppf
      "====================== <OPTIONS> \
       ======================@.%a@.===================== </OPTIONS> \
       ======================@.======================= <PARSED> \
       =======================@.%a@.====================== </PARSED> \
       =======================@.@.@."
      Pb_option.pp_set options
      (Format.pp_print_option Validate.pp_field_rules)
      parsed;
    ()
  )

let run proto =
  let protos =
    Pb_parsing.parse_file
      (fun f ->
        match f with
        | "test.proto" -> f, proto
        | _ -> f, "")
      "test.proto"
  in
  let ppf = Format.std_formatter in
  Format.set_margin 150;
  protos
  |> List.iter (fun proto ->
         proto.Pt.messages
         |> List.iter (fun message ->
                message.Pt.message_body
                |> List.iter (function
                     | Pt.Message_field field -> process_field_options ppf field
                     | _ -> ())))

let test_cases =
  [
    {|
      syntax = "proto3";
      
      package examplepb;
      
      import "validate/validate.proto";
      
      message Person {
        uint64 id = 1 [(validate.rules).uint64.gt = 999];
      
        string email = 2 [(validate.rules).string.email = true];
      
        string name = 3 [(validate.rules).string = {
          pattern:   "^[^[0-9]A-Za-z]+( [^[0-9]A-Za-z]+)*$",
          max_bytes: 256,
        }];
      
        Location home = 4 [(validate.rules).message.required = true];
      
        message Location {
          double lat = 1 [(validate.rules).double = {gte: -90,  lte: 90}];
          double lng = 2 [(validate.rules).double = {gte: -180, lte: 180}];
        }
      }
    |};
  ]

let () = List.iter run test_cases
