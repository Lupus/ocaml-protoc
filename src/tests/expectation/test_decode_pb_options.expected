====================== <OPTIONS> ======================
[{
   "(validate.rules)": {"uint64": {"gt": 999}}
   }]
===================== </OPTIONS> ======================
======================= <PARSED> =======================
{ message = None;
  type_ = Some(Uint64({ const = None; lt = None; lte = None; gt = Some(999); gte = None; in_ = []; not_in = []; ignore_empty = None; }));
}
====================== </PARSED> =======================


====================== <OPTIONS> ======================
[{
   "(validate.rules)": {"string": {"email": true}}
   }]
===================== </OPTIONS> ======================
======================= <PARSED> =======================
{ message = None;
  type_ =
    Some(
      String(
        { const = None;
          len = None;
          min_len = None;
          max_len = None;
          len_bytes = None;
          min_bytes = None;
          max_bytes = None;
          pattern = None;
          prefix = None;
          suffix = None;
          contains = None;
          not_contains = None;
          in_ = [];
          not_in = [];
          well_known = Some(Email(true));
          strict = Some(true);
          ignore_empty = None;
        }));
}
====================== </PARSED> =======================


====================== <OPTIONS> ======================
[{
   "(validate.rules)": {"string": {"pattern": "^[^[0-9]A-Za-z]+( [^[0-9]A-Za-z]+)*$",
                                   "max_bytes": 256}}
   }]
===================== </OPTIONS> ======================
======================= <PARSED> =======================
{ message = None;
  type_ =
    Some(
      String(
        { const = None;
          len = None;
          min_len = None;
          max_len = None;
          len_bytes = None;
          min_bytes = None;
          max_bytes = Some(256);
          pattern = Some("^[^[0-9]A-Za-z]+( [^[0-9]A-Za-z]+)*$");
          prefix = None;
          suffix = None;
          contains = None;
          not_contains = None;
          in_ = [];
          not_in = [];
          well_known = None;
          strict = Some(true);
          ignore_empty = None;
        }));
}
====================== </PARSED> =======================


====================== <OPTIONS> ======================
[{
   "(validate.rules)": {"message": {"required": true}}
   }]
===================== </OPTIONS> ======================
======================= <PARSED> =======================
{ message = Some({ skip = None; required = Some(true); }); type_ = None; }
====================== </PARSED> =======================


